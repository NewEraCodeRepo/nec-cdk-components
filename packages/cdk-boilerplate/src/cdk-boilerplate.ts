#!/usr/bin/env node
import * as cloudtrail from '@aws-cdk/aws-cloudtrail';
import { Bucket } from '@aws-cdk/aws-s3';
import { Construct } from '@aws-cdk/core';

export interface ICDKBoilerplateProps {
  bucketName: string;
}

/**
 * Boilerplate
 */
export class CDKBoilerplate extends Construct {
  public bucketWebsiteUrl: string;

  constructor(scope: Construct, id: string, props: ICDKBoilerplateProps) {
    super(scope, id);

    const bucket = new Bucket(this, 'CloudTrailBucket', {
      bucketName: props.bucketName,
    });
    // tslint:disable-next-line:no-unused-expression
    new cloudtrail.Trail(this, 'CloudTrail', {
      bucket,
      isMultiRegionTrail: true,
    });

    this.bucketWebsiteUrl = bucket.bucketWebsiteUrl;
  }
}
