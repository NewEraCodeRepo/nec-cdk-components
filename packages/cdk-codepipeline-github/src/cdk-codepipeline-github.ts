import { IStage, Pipeline } from '@aws-cdk/aws-codepipeline';
import { LambdaFunction } from '@aws-cdk/aws-events-targets';
import { PolicyStatement } from '@aws-cdk/aws-iam';
import { Code, Function, Runtime } from '@aws-cdk/aws-lambda';
import { Construct, Duration, SecretValue } from '@aws-cdk/core';
import { join } from 'path';

export interface ICdkCodepipelineGithubProps {
  /**
   * The Codepipeline stage publishing status events
   */
  readonly stage?: IStage;

  /**
   * The Codepipeline publishing status events
   */
  readonly pipeline?: Pipeline;

  /**
   * The arn of a secret in the Secrets Manager with
   * a github token value.
   */
  readonly gitHubSecretArn: string;
}

/**
 * Publishes status events from a stage in Codepipeline
 * to a Github repo using webhooks.
 *
 * SAM template conversion by Julia Jacobs
 * {@see {@link https://github.com/jewelsjacobs}}
 *
 * SAM template created by Jens Eickmeyer
 * {@see {@link https://github.com/jenseickmeyer/github-commit-status-bot}}
 * {@see {@link https://scratchpad.blog/devops/howto/send-commit-status-from-codepipeline-to-github/}}
 *
 * @example <caption>Example usage of construct in pipeline.</caption>
 * // One possible way to get an existing secret's arn - using AwsCustomResource
 * // and AWS SecretsManager SDK
 * const getGithubSecret = new AwsCustomResource(this, 'GetGithubSecret', {
 *   onUpdate: {
 *      service: 'SecretsManager',
 *      action: 'describeSecret',
 *      parameters: {
 *         SecretId: 'TOKENS/PIPELINE/GithubToken' // ID of secret
 *      }
 *   }
 * });
 * // ARN is a property in the describeSecret SDK response
 * const gitHubSecretArn = `${getGithubSecret.getData("ARN")}` as string;
 *
 * // pipeline
 * const pipeline = new codepipeline.Pipeline(this, 'Pipeline', {
 *   restartExecutionOnUpdate: true
 * });
 *
 * // source stage
 * pipeline.addStage({
 *   stageName: 'Source',
 *   actions: [sourceAction]
 * });
 *
 * // build stage
 * const buildStage = pipeline.addStage({
 *   stageName: 'Build',
 *   actions: [buildAction]
 * });
 *
 * // Github status checks based on status of the build stage
 * new CdkCodepipelineGithub(this, 'GithubCodepipeline', {
 *   stage: buildStage,
 *   gitHubSecretArn
 * });
 *
 * // deploy stage
 * pipeline.addStage({
 *   stageName: 'Deploy',
 *   actions: [deployAction]
 * });
 *
 * // or check status of pipeline itself:
 * new CdkCodepipelineGithub(this, 'GithubCodepipeline', {
 *   pipeline,
 *   gitHubSecretArn
 * });
 */
export class CdkCodepipelineGithub extends Construct {
  constructor(scope: Construct, id: string, props: ICdkCodepipelineGithubProps) {
    super(scope, id);

    const { stage, pipeline, gitHubSecretArn } = props;

    const statusLambda = new Function(this, 'StatusLambda', {
      code: Code.asset(join(__dirname, '..', 'lambda_packages', 'package.zip')),
      environment: {
        ACCESS_TOKEN: SecretValue.secretsManager(gitHubSecretArn).toString(), // github token passed to Lambda
      },
      handler: 'index.handler',
      runtime: Runtime.NODEJS_10_X,
      timeout: Duration.seconds(300),
    });

    const lambdaTarget = new LambdaFunction(statusLambda);

    if (stage) {
      stage.onStateChange('StageOnStateChange', lambdaTarget, {
        eventPattern: {
          detail: {
            state: ['STARTED', 'SUCCEEDED', 'FAILED'],
          },
        },
      });
    }

    if (pipeline) {
      pipeline.onStateChange('StageOnStateChange', {
        eventPattern: {
          detail: {
            state: ['STARTED', 'SUCCEEDED', 'FAILED'],
          },
        },
        target: lambdaTarget,
      });
    }

    if (!stage && !pipeline) {
      throw Error('Choose a pipeline or stage to send event status messages from');
    }

    statusLambda.addToRolePolicy(
      new PolicyStatement({
        actions: ['*'],
        resources: ['*'],
      }),
    );
  }
}
