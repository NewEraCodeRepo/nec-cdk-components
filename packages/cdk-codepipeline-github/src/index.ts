export * from './cdk-codepipeline-github';
