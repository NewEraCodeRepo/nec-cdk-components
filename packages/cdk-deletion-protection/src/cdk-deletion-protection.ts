#!/usr/bin/env node
import { CfnDeletionPolicy, CfnResource, Construct } from '@aws-cdk/core';
import { AwsCustomResource } from '@aws-cdk/custom-resources';

export interface ICDKDeletionProtectionProps {
  resourceDeletionPolicy?: string; // "DELETE" | "RETAIN" | "SNAPSHOT"
  terminationProtection?: boolean;
}

/**
 * DeletionProtection
 *
 * Example usage:
 *
 * const deletionProtection = new CDKDeletionProtection(this, 'ProtectionPolicies', {
 *    resourceDeletionPolicy: "RETAIN",
 *    terminationProtection: true,
 * });
 *
 * // enable stack termination deletion protection:
 * deletionProtection.setStackTerminationProtection(this.stackName);
 *
 * // protect a Cognito User Pool from getting deleted on update:
 * const userPool = new UserPool(this, 'cdkUserPool', {...});
 * const cfnUserPool = userPool.node.defaultChild as CfnUserPool;
 * deletionProtection.setResourcePolicy(cfnUserPool);
 */
export class CDKDeletionProtection extends Construct {
  public setResourcePolicy: (cfnResource: CfnResource) => void;
  public setStackTerminationProtection: (stackName: string) => void;
  constructor(scope: Construct, id: string, props?: ICDKDeletionProtectionProps) {
    super(scope, id);

    this.setResourcePolicy = (cfnResource: CfnResource) => {
      if (props && props.hasOwnProperty('resourceDeletionPolicy')) {
        switch (props.resourceDeletionPolicy) {
          case 'DELETE':
            cfnResource.cfnOptions.updateReplacePolicy = CfnDeletionPolicy.DELETE;
            cfnResource.cfnOptions.deletionPolicy = CfnDeletionPolicy.DELETE;
            break;
          case 'RETAIN':
            cfnResource.cfnOptions.updateReplacePolicy = CfnDeletionPolicy.RETAIN;
            cfnResource.cfnOptions.deletionPolicy = CfnDeletionPolicy.RETAIN;
            break;
          case 'SNAPSHOT':
            cfnResource.cfnOptions.updateReplacePolicy = CfnDeletionPolicy.SNAPSHOT;
            cfnResource.cfnOptions.deletionPolicy = CfnDeletionPolicy.SNAPSHOT;
            break;
          default:
            cfnResource.cfnOptions.updateReplacePolicy = CfnDeletionPolicy.RETAIN;
            cfnResource.cfnOptions.deletionPolicy = CfnDeletionPolicy.RETAIN;
            break;
        }
      } else {
        throw new Error('Resource deletion policy required.');
      }
    };

    this.setStackTerminationProtection = (stackName: string) => {
      if (props && props.hasOwnProperty('terminationProtection')) {
        // Enable stack termination protection
        // using the AWS SDK CloudFormation updateTerminationProtection API:
        // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CloudFormation.html#updateTerminationProtection-property

        // tslint:disable-next-line:no-unused-expression
        new AwsCustomResource(this, 'UpdateTerminationProtection', {
          onUpdate: {
            action: 'updateTerminationProtection',
            parameters: {
              EnableTerminationProtection: props.terminationProtection,
              StackName: stackName,
            },
            service: 'CloudFormation',
          },
        });
      } else {
        throw Error('Termination protection option must be set to true or false.');
      }
    };
  }
}
