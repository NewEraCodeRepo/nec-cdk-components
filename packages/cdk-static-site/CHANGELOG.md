# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.8.3](https://github.com/NewEraCodeRepo/nec-cdk-components/compare/v1.8.2...v1.8.3) (2020-01-10)

**Note:** Version bump only for package @NewEraCode/cdk-static-site





## [1.8.2](https://github.com/NewEraCodeRepo/nec-cdk-components/compare/v1.8.1...v1.8.2) (2020-01-10)

**Note:** Version bump only for package @NewEraCode/cdk-static-site





## [1.8.1](https://github.com/NewEraCodeRepo/nec-cdk-components/compare/v1.8.0...v1.8.1) (2019-10-23)

**Note:** Version bump only for package @NewEraCode/cdk-static-site





# 1.8.0 (2019-10-23)


### Bug Fixes

* merge fixes ([d45bfb0](https://github.com/NewEraCodeRepo/nec-cdk-components/commit/d45bfb04a6cab25f96ac5e3ee895aa72cd223139))
* merge fixes ([4799bd6](https://github.com/NewEraCodeRepo/nec-cdk-components/commit/4799bd6df82839b43e19d40be99f378352d05c74))
* updated mods and snapshots ([0f3362a](https://github.com/NewEraCodeRepo/nec-cdk-components/commit/0f3362ac7a298bd2d6208e12a6faefabd6eb1dfb))


### Features

* added cdk-deletion-protection source code ([5701c1a](https://github.com/NewEraCodeRepo/nec-cdk-components/commit/5701c1a191ad2bfc13c506b94f8e1096287965ac))
* update and add new mods ([0a21999](https://github.com/NewEraCodeRepo/nec-cdk-components/commit/0a21999e9ade9130b52eacbabae9e2d635436987))
* update node mods ([28fd58b](https://github.com/NewEraCodeRepo/nec-cdk-components/commit/28fd58be5049d474db9dd60a308e03d834bfd8f0))





## [1.7.6](https://github.com/NewEraCodeRepo/nec-cdk-components/compare/v1.7.5...v1.7.6) (2019-10-22)

**Note:** Version bump only for package @NewEraCode/cdk-static-site





## [1.7.5](https://github.com/NewEraCodeRepo/nec-cdk-components/compare/v1.7.2...v1.7.5) (2019-10-22)


### Bug Fixes

* merge fixes ([4799bd6](https://github.com/NewEraCodeRepo/nec-cdk-components/commit/4799bd6df82839b43e19d40be99f378352d05c74))
