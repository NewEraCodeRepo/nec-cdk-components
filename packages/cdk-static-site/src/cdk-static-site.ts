#!/usr/bin/env node
/* tslint:disable */
import { CloudFrontWebDistribution, SecurityPolicyProtocol, SSLMethod } from '@aws-cdk/aws-cloudfront';
import { HostedZone, ARecord, RecordTarget } from '@aws-cdk/aws-route53';
import { CloudFrontTarget } from '@aws-cdk/aws-route53-targets/lib';
import { Bucket } from '@aws-cdk/aws-s3';
import { BucketDeployment, Source } from '@aws-cdk/aws-s3-deployment';
import { Construct, RemovalPolicy } from '@aws-cdk/core';
import * as path from 'path';

export interface CdkStaticSiteProps {
  bucketConfiguration: {
    bucketName: string;
    removalPolicy: RemovalPolicy;
    websiteIndexDocument: string;
    websiteErrorDocument: string;
    source: string;
  };
  aliasConfiguration: {
    domainName: string;
    siteDomain: string;
    acmCertRef: string;
  };
}

/**
 * Static site infrastructure, using an S3 bucket for the content.
 *
 * The site redirects from HTTP to HTTPS, using a CloudFront distribution,
 * Route53 alias record, and ACM certificate.
 *
 * The ACM certificate is expected to be created and validated outside of the CDK,
 * with the certificate ARN stored in an SSM Parameter.
 */
export class CdkStaticSite extends Construct {
  constructor(parent: Construct, name: string, props: CdkStaticSiteProps) {
    super(parent, name);

    const { bucketName, removalPolicy, websiteIndexDocument, websiteErrorDocument, source } = props.bucketConfiguration;
    const { domainName, siteDomain, acmCertRef } = props.aliasConfiguration;

    // Content bucket
    const siteBucket = new Bucket(this, 'SiteBucket', {
      bucketName,
      publicReadAccess: true,
      // The default removal policy is RETAIN, which means that cdk destroy will not attempt to delete
      // the new bucket, and it will remain in your account until manually deleted. By setting the policy to
      // DESTROY, cdk destroy will attempt to delete the bucket, but will error if the bucket is not empty.
      removalPolicy, // DESTROY NOT recommended for production code
      websiteErrorDocument,
      websiteIndexDocument,
    });

    // CloudFront distribution that provides HTTPS
    const distribution = new CloudFrontWebDistribution(this, 'SiteDistribution', {
      aliasConfiguration: {
        acmCertRef,
        names: [siteDomain],
        securityPolicy: SecurityPolicyProtocol.TLS_V1_1_2016,
        sslMethod: SSLMethod.SNI,
      },
      defaultRootObject: websiteIndexDocument,
      originConfigs: [
        {
          behaviors: [{ isDefaultBehavior: true }],
          s3OriginSource: {
            s3BucketSource: siteBucket,
          },
        },
      ],
    });

    const placeHolderSource = path.join(__dirname, '..', 'website');

    new BucketDeployment(this, 'DeployWebsite', {
      destinationBucket: siteBucket,
      distribution,
      retainOnDelete: removalPolicy === RemovalPolicy.RETAIN,
      sources: [Source.asset(source || placeHolderSource)],
    });

    // Route53 alias record for the CloudFront distribution
    const zone = HostedZone.fromLookup(this, 'Zone', {
      domainName,
    });
    new ARecord(this, 'SiteAliasRecord', {
      recordName: siteDomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
      zone,
    });
  }
}
