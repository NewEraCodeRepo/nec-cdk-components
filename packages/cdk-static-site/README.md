# NewEraCode Static Website Component

## Description

An [AWS CDK](https://docs.aws.amazon.com/cdk/latest/guide/home.html) construct that creates static site infrastructure, which uses an S3 bucket for the content.

The site redirects from HTTP to HTTPS, using a CloudFront distribution,
Route53 alias record, and ACM certificate.

The ACM certificate is expected to be created and validated outside of the CDK,
with the certificate ARN stored in an SSM Parameter.

## Usage

### Install

```bash
npm i @NewEraCode/cdk-static-website
```
### Import

**NodeJS**
```javascript
const cds = require('@NewEraCode/cdk-static-site');
new cds.CdkStaticSite(....)
```
**TypeScript / Babel**
```typescript
import { CdkStaticSite } from '@NewEraCode/cdk-static-site';
```

Example:

```typescript
import { App, RemovalPolicy, Stack, StackProps } from '@aws-cdk/core';
import { CdkStaticSite } from './cdk-static-site';

export class ComponentStack extends Stack {
    public constructor(parent: App, name: string, props?: StackProps) {
        super(parent, name, props);

        new CdkStaticSite(this, 'StaticWebsite', {
            bucketConfiguration: {
                bucketName: "myBucket",
                removalPolicy: RemovalPolicy.RETAIN,
                websiteIndexDocument: "index.html",
                websiteErrorDocument: "error.html",
                source: "websiteFiles" // local directory with static files
            },
            aliasConfiguration: {
                domainName: "domain.com"
                siteDomain: "example.domain.com",
                acmCertRef: "arn:aws:acm:us-east-1:....."
            },
        });
    }
}
```
