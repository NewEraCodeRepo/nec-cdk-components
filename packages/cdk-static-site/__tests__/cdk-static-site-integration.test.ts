import { App, Environment, RemovalPolicy, Stack } from '@aws-cdk/core';
import 'jest-cdk-snapshot';
import mock from 'mock-fs';
import { join } from 'path';
// tslint:disable-next-line: no-var-requires
import { CdkStaticSite } from '../src/cdk-static-site';
jest.mock('@aws-cdk/aws-s3-deployment');

const dirPath = join(__dirname, 'websiteFiles');

describe('cdk-static-site', () => {
  beforeEach(() => {
    mock(
      {
        dirPath: mock.directory({
          items: {
            'error.html': '<html><body><h1>Bye!</h1></body></html>',
            'index.html': '<html><body><h1>Hi!</h1></body></html>',
          },
          mode: '0755',
        }),
      },
      { createCwd: false, createTmp: false },
    );
  });

  test('snapshot', () => {
    const env: Environment = {
      account: '123456789012',
      region: 'us-east-1',
    };

    const app = new App();
    const stack = new Stack(app, 'Test', { env });

    // tslint:disable-next-line: no-unused-expression
    new CdkStaticSite(stack, 'StaticWebsite', {
      aliasConfiguration: {
        acmCertRef: 'arn:aws:acm:us-east-1:123456789012:certificate/955290e7-e7e5-47b1-9077-b84e7dc1747a',
        domainName: 'domain.com',
        siteDomain: 'example.domain.com',
      },
      bucketConfiguration: {
        bucketName: 'test-bucket',
        removalPolicy: RemovalPolicy.RETAIN,
        source: dirPath,
        websiteErrorDocument: 'error.html',
        websiteIndexDocument: 'index.html',
      },
    });
    mock.restore();
    expect(stack).toMatchCdkSnapshot();
  });
});
