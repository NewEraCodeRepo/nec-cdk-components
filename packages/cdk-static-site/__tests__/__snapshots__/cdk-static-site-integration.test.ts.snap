// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cdk-static-site snapshot 1`] = `
Object {
  "Resources": Object {
    "StaticWebsiteSiteAliasRecord5B5BAA80": Object {
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": Object {
            "Fn::GetAtt": Array [
              "StaticWebsiteSiteDistributionCFDistributionCF168A59",
              "DomainName",
            ],
          },
          "HostedZoneId": "Z2FDTNDATAQYW2",
        },
        "HostedZoneId": "DUMMY",
        "Name": "example.domain.com.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "StaticWebsiteSiteBucket57CB7A59": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketName": "test-bucket",
        "WebsiteConfiguration": Object {
          "ErrorDocument": "error.html",
          "IndexDocument": "index.html",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "StaticWebsiteSiteBucketPolicyF2353F67": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "StaticWebsiteSiteBucket57CB7A59",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": "*",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "StaticWebsiteSiteBucket57CB7A59",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "StaticWebsiteSiteDistributionCFDistributionCF168A59": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "Aliases": Array [
            "example.domain.com",
          ],
          "DefaultCacheBehavior": Object {
            "AllowedMethods": Array [
              "GET",
              "HEAD",
            ],
            "CachedMethods": Array [
              "GET",
              "HEAD",
            ],
            "Compress": true,
            "ForwardedValues": Object {
              "Cookies": Object {
                "Forward": "none",
              },
              "QueryString": false,
            },
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "StaticWebsiteSiteBucket57CB7A59",
                  "RegionalDomainName",
                ],
              },
              "Id": "origin1",
              "S3OriginConfig": Object {},
            },
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": Object {
            "AcmCertificateArn": "arn:aws:acm:us-east-1:123456789012:certificate/955290e7-e7e5-47b1-9077-b84e7dc1747a",
            "MinimumProtocolVersion": "TLSv1.1_2016",
            "SslSupportMethod": "sni-only",
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
  },
}
`;
